woocommerce/qit-cli
# Fix PHP 8.4 deprecation

This PR:

- Updates DI52 to ^4, which [fixes](https://github.com/lucatume/di52/pull/66) the PHP 8.4 deprecation
- Adds a workflow to run our PHPUnit and Phan in all versions we support (7.2-8.4)
- Make some small tweaks to our tests so that they run on PHP 7.2
- This workflow will run on PR pushes

### Testing Instructions
- Run `make check-php-versions`, assert that it builds a docker image for each PHP version
- Assert that each docker image has the expected PHP version
- Then run `make tests-all` and assert it runs the PHPUnit and Phan in each of the supported versions, and that it passes on all
- Double-check the PHPUnit tests on CI in this PR

There's only one Deprecated being triggered in the test itself in PHP 8.4, but it's safe to ignore:

```
Deprecated: Spatie\Snapshots\MatchesSnapshots::assertMatchesSnapshot(): Implicitly marking parameter $driver as nullable is deprecated, the explicit nullable type must be used instead in /app/src/vendor/spatie/phpunit-snapshot-assertions/src/MatchesSnapshots.php on line 48
```

We currently use on an older version of `spatie/phpunit-snapshot-assertions` that supports PHP 7.2. However, this version triggers a deprecation warning when running on PHP 8.4. While upgrading to the latest release of the package would resolve the issue, it requires a minimum of PHP 8.1, which would break compatibility with older environments.

Given these constraints, the most practical solution is to simply ignore this deprecation notice under PHP 8.4. Since the warning appears only during tests.
